### PR TITLE
Problem (Fix #1040): No reward distribution integration tests

### LINKED ISSUE #1040: Problem: No reward distribution integration tests
- extra assertions in existing tests -- e.g. jailed validators don't get rewarded ?
- some expectation on amounts when rewarded -- e.g. with bonus slash? (maybe lower bound if block generation is hard to control)

### PR DESCRIPTION
**Pull request by yihuang**
Solution:
    - Add a dedicated reward integration test
      Check:
      - monetary expansion formula, tau decaying
      - no reward for jailed node
      - slashing is added to reward pool

Based on https://github.com/crypto-com/chain/pull/1466, rebase after that is merged.

### CODE REVIEW

**Comment by tomtau** [comment]
bors try

**Comment by yihuang** [comment]
> ## try
> Build failed:
> 
> * [drone-hw/push](http://13.82.172.109/crypto-com/chain/460)

Should be fixed now.

**Comment by tomtau** [comment]
bors retry

**Comment by tomtau** [comment]
bors retry

**Comment by yihuang** [comment]
> looks ok. I wonder if the account fetching could just be done with the existing "account" endpoint + https://github.com/polkascan/py-scale-codec -- I guess one would just need to add that json type mapping file for staking state etc.?

It should work in that way too.
But still, it'll require all the client languages do the scale-codec decoding, and from what I see, using the JSON format doesn't have a downside.

**Comment by yihuang** [comment]
> can be rebased now?

Done

**Comment by yihuang** [comment]
> looks ok. I wonder if the account fetching could just be done with the existing "account" endpoint + https://github.com/polkascan/py-scale-codec -- I guess one would just need to add that json type mapping file for staking state etc.?

The py-scale-codec will be brought in when fixing #1313.

**Comment by tomtau** [comment]
bors try

File: chain-abci/src/app/query.rs [hunk 1]
```diff
@@ -191,21 +191,17 @@ impl<T: EnclaveProxy> ChainNodeApp<T> {
...
-                    resp.value = serde_json::to_string(&(
-                        maccount,
-                        if _req.prove { Some(proof) } else { None },
-                    ))
-                    .unwrap()
-                    .into_bytes();
+                    let (mstaking, proof) = get_with_proof(&self.storage, version, &address);
+                    let mproof = if _req.prove { Some(proof) } else { None };
+                    resp.value =
+                        [mstaking.encode().as_slice(), mproof.encode().as_slice()].concat();
```
[the remark this answers is not in the document]
Comment (yihuang) [reply chain-abci/src/app/query.rs]: Done

**Comment by yihuang** [comment]
> ABCI query response has a dedicated "proof" field that should be filled in if the request has "prove" set to true -- here the proof is always bundled in the value response.
> Following the standard behaviour will then make it easier to adapt e.g. for https://github.com/cosmos/ics/blob/master/spec/ics-023-vector-commitments/README.md https://github.com/informalsystems/ibc-rs/blob/master/relayer/relay/src/query.rs#L32

Done, add it into the proof field.

**Pull request merged**